### PR TITLE
Fix amp-bind in PWA

### LINKED ISSUE #9916: Issue using amp-bind with Shadow DOM
I'm having problems using amp-bind within a PWA that uses the Shadow DOM.

Steps to reproduce:

1. Go to the example I've created at: https://www.edhollinghurst.com/amp-bind-pwa/amp-bind-pwa.html
2. Observe that if you click an answer it is correctly bound to the element used to display the answer.
3. Refresh the page. This should move you across to the 'PWA' with the 'App Shell' header displayed.
4. Trying clicking an answer. Observe that it is no longer bound to the element used to display the answer.

@choumx FYI

### PR DESCRIPTION
**Pull request by choumx**
Fixes #9916.

- Scan `ampdoc.getBody()`, not `ampdoc.win.document.body` (doesn't work for shadow docs)
- Do *not* fallback to top window when retrieving the Bind element service (instead, fallback to ampdoc since it's a doc-scope service)
- Add relevant tests for `element-service.js` and `bind-impl.js`

This PR also requires passing a boolean `opt_fallbackToTopWin` to allow crossing FIE boundary when getting services. I think this is a safer default, e.g. for A4A.

### CODE REVIEW

File: src/service.js [hunk 1]
```diff
@@ -68,46 +74,37 @@ export class EmbeddableService {
...
- * @return {?Object} The service.
- */
-export function getExistingServiceOrNull(win, id) {
-  win = getTopWindow(win);
-  if (isServiceRegistered(win, id)) {
-    return getServiceInternal(win, id);
-  } else {
-    return null;
-  }
-}
```
Comment (choumx) [inline src/service.js]: This was just moved below for consistent ordering.

File: src/service.js [hunk 2]
```diff
@@ -161,21 +156,6 @@ function getLocalExistingServiceForEmbedWinOrNull(embedWin, id) {
...
- * `vsyncFor(win)`) for type safety and because the factory should not be
- * passed around.
- * @param {!Window} win
- * @param {string} id of the service.
- * @template T
- * @return {T}
- */
-export function getService(win, id) {
-  win = getTopWindow(win);
-  return getServiceInternal(win, id);
```
Comment (choumx) [inline src/service.js]: This was just moved below for consistent ordering.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 3]
```diff
@@ -135,15 +136,21 @@ export class Bind {
     /** @const @private {!../../../src/service/viewer-impl.Viewer} */
     this.viewer_ = viewerForDoc(this.ampdoc);
 
-    /**
+    const bodyReadyPromise = (opt_win)
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: Doesn't this resolve to a `body`?

File: src/element-service.js [hunk 4]
```diff
@@ -173,10 +174,16 @@ export function getElementServiceIfAvailableForDocInEmbedScope(
...
     return /** @type {!Promise<?Object>} */ (Promise.resolve(s));
   }
+  // Return embed-scope element service promise if scheduled.
   if (nodeOrDoc.nodeType) {
+    // In embeds, doc-scope services are window-scope.
     const win = /** @type {!Document} */ (
         nodeOrDoc.ownerDocument || nodeOrDoc).defaultView;
-    return elementServicePromiseOrNull(win, id, extension);
+    const topWin = getTopWindow(win);
+    // Don't return promise unless this is definitely FIE to avoid covfefe.
```
Comment (jridgewell) [inline src/element-service.js]: I don't understand this.

Comment (choumx) [inline src/element-service.js]: Made this comment clearer. The current system is pretty confusing and I'm planning on refactoring this soon.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 3 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: `whenBodyAvailable` does, but `waitForBodyPromise` doesn't. Changed it to use resolve value but not sure it's cleaner.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 5]
```diff
@@ -135,17 +136,19 @@ export class Bind {
     /** @const @private {!../../../src/service/viewer-impl.Viewer} */
     this.viewer_ = viewerForDoc(this.ampdoc);
 
-    /**
+    const bodyPromise = (opt_win)
+        ? waitForBodyPromise(opt_win.document)
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: You should make it resolve to a `body`. 😉

File: extensions/amp-bind/0.1/bind-impl.js [hunk 6]
```diff
@@ -135,17 +136,19 @@ export class Bind {
...
      */
-    this.initializePromise_ = Promise.all([
-      waitForBodyPromise(this.localWin_.document), // Wait for body.
-      this.viewer_.whenFirstVisible(), // Don't initialize in prerender mode.
-    ]).then(() => {
-      const rootNode = dev().assertElement(this.localWin_.document.body);
-      return this.initialize_(rootNode);
-    });
+    this.initializePromise_ =
+        this.viewer_.whenFirstVisible().then(bodyPromise).then(body => {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: the `then(bodyPromise)` does not return a Promise, it returns `whenFirstVisible`'s again. You have to pass in a function.

File: src/element-service.js [hunk 7]
```diff
@@ -173,10 +174,17 @@ export function getElementServiceIfAvailableForDocInEmbedScope(
   if (s) {
     return /** @type {!Promise<?Object>} */ (Promise.resolve(s));
   }
+  // Return embed-scope element service promise if scheduled.
   if (nodeOrDoc.nodeType) {
     const win = /** @type {!Document} */ (
         nodeOrDoc.ownerDocument || nodeOrDoc).defaultView;
-    return elementServicePromiseOrNull(win, id, extension);
+    const topWin = getTopWindow(win);
+    // In embeds, doc-scope services are window-scope. But make sure to
```
Comment (jridgewell) [inline src/element-service.js]: I still don't understand it, but ok.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 6 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Oops, done.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 5 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: 😩  How about in fix-it next week?

**Pull request merged**